release: Prepare for MacOS builds on GitHub Actions

camscale opened the pull request:
Update the `Makefile` and `build.assets/build-common.sh` to allow
signing keys to be parameterized for the signing and notarization
process, used for making releases for MacOS.

This moves some variables from the `build-common.sh` shell script to the
Makefile, which then exports those variables in the recipies that run
the scripts.

The hard-coded key IDs have been replaced with querying the keyring
based on the name of the key. This allows us to use different keys in
Drone and GitHub Actions as requested by the security team. It also
makes rotating the keys easier as no repo changes are needed to use the
new keys.

The default IDs and other signing parameters are for the production
environment (call `promote` in GitHub) as Drone has no concept of this
environment config. Developers can manually export
`ENVIRONMENT_NAME=build` to use the development keys locally. When
developing the GitHub Actions workflows in the enterprise repository,
changing the `environment: promote` to `environment: build` will
automatically pick up the correct environment name and use the dev keys.

Issue: https://github.com/gravitational/teleport/issues/20319

camscale committed:
release: Move Mac signing vars from script to Makefile

Move the variables for Mac signing from the `build-common.sh` shell
script to the `Makefile`. These vars will need to be passed to other
build processes to parameterize the signing for different GitHub Actions
build environments.

The switch on `ENVIRONMENT_NAME` allows different secrets to be
available in GitHub Actions for production (promote) vs developer
(build) builds. The default environment name is `promote` so as to be
compatible with the existing Drone setup, which does not define
`ENVIRONMENT_NAME`.

camscale committed:
release: Determine Mac signing key IDs automatically

Remove the hard-coded MacOS signing key IDs from the Makefile and find
them dynamically based on the name of the key. This allows GitHub
Actions to be set up with new keys different to the ones on the Drone
builders. As long as we keep the same name on the keys, we can rotate
the keys without needing to update the IDs in the Makefile.

This requires us to be more judicious about exporting the variables as
exporting them causes them to be evaluated. We do not want to evaluate
them on non-darwin targets, and on darwin, we should only evaluate it if
needed for a recipe. So use a dynamic `eval` in the recipes that need
the environment variables.

camscale committed:
release: Pass key & team ID to notarize tool

Override the hard-coded values in `notarize-apple-binaries` and pass the
values we get based on the GitHub Actions environment. This allows us to
sign and notarize software in a development branch more easily when
working on the signing and notarizing process. This will not happen
automatically, but it is expected that a developer can manually trigger
a workflow to perform building, signing and notarizing from a dev
branch where the workflow has temporarily changed the environment to
`build`.

A similar change to the `Makefile` in the teleport.e repository goes
with this change.

This adds a new bundle ID of `com.goteleport.dev` for the dev build of
Teleport. This follows the same pattern as used for the dev build of the
`tsh` binary and the current production bundle ID for Teleport.
Previously there was no dev signing/notarizing process for the set of
Teleport binaries.

camscale committed:
release: Add script to setup the MacOS keychain for signing

Add a script for setting up the MacOS keychain for signing applications
and packages. It encapsulates the `security` commands to add either or
both application keys and installer keys. The keys can be either
base64-encoded in environment variables, or `.p12` files on disk, making
it useful for local development.

camscale commented:
@codingllama Maybe RTYI, as it reworks the key and other parameter handling used in the mac build scripts you wrote. Would be happy to have your feedback too.

r0mant commented on Makefile:
```diff
@@ -429,6 +429,34 @@ build-archive:
 
+# Set MacOS/Darwin cert vars for signing/notarizing
+# TSH_SKELETON is a directory name relative to build.assets/macos/
+ENVIRONMENT_NAME ?= promote
```
Can all this new stuff be moved into a separate .mk file (maybe `darwin.mk` or something) and included here?

camscale replied:
I was considering that, but wasn't sure where to put it. I could put it at the top-level, but that's a bit polluting. We do have `common.mk` there already though. I could put it in `build.assets/` but that already has a `Makefile` so it may look like it is part of that `Makefile` not the top-level one. I could create a new top-level directory `mk/` and put it there and even move the others there. But that's more churn than I want right now.

So in the absence of a winning idea, I just put it in the top-level `Makefile` :-|

Do you have a preference?

camscale replied:
I went with `darwin-signing.mk` as I felt that `darwin.mk` should have all the darwin stuff in it. Happy to rename it to the shorter name if you prefer.

r0mant commented on Makefile:
```diff
@@ -429,6 +429,34 @@ build-archive:
+ENVIRONMENT_NAME ?= promote
+
+# TODO(camh): Verify key names for production keys
+DEVELOPER_KEY_NAME_promote = Developer ID Application: Ada Lin
```
Is "Ada Lin" expected here and below?

r0mant replied:
Also, dding comments on all these variables to explain what they are for would be helpful for future maintainers.

wadells replied:
The production promote/release certificate info can be found here:

https://goteleport.com/security/

It sounds like we're ready to have those secrets configured in https://github.com/gravitational/teleport.e/issues/832, right @camscale?

camscale replied:
dding? 

I'm not sure what the names for the promote keys are yet. @wadells Do you have these set up in the `promote` environment yet? Can you let me know?

camscale replied:
Yep, ready for the `promote` environment now/shortly. This all works with the `build` environment. I expect I'll merge with `build` and land a couple more PRs on top, and then switch to promote and ensure the end-to-end release process works with that.

r0mant commented on Makefile:
```diff
@@ -1091,6 +1122,7 @@ endif
 # build .pkg
 .PHONY: pkg
 pkg:
+	$(eval export DEVELOPER_ID_APPLICATION DEVELOPER_ID_INSTALLER TEAMID TSH_BUNDLEID TSH_SKELETON)
```
Why are these evals needed? The environment variables are being set above, isn't it enough? We don't use eval anywhere else in our makefiles.

camscale replied:
They need to be exported. If we unconditionally export them, then the `$(shell security ...)` command will be run every time you invoke `make`. If they go into a `ifeq ($OS),darwin)` block, they will run every time you run `make` on darwin even if you are not making a release or running anything that needs the key IDs. This might be ok, but I was concerned about the side-effects and affecting developers doing normal builds.

Doing it this way only exports the variables if those targets are invoked which means the `security` command is only run then.

BUT, I just remembered there is an unconditional `export` further up in the Makefile - I need to check to see if that exports only the things before that or everything. I didn't notice the command trying to run on Linux, but I will double-check that.

BTW I forgot to mention in the PR description that the individual commits stand on their own and have commit messages describing them. I think I mention some of this in the relevant commit message.

camscale replied:
I will comment this in the Makefile

camscale replied:
It looks like this does not work on Linux due to the top-level `export` directive. I will need to add an unexport for those names so it does not unconditionally run the `security find-identity` command to look up the key id. But I will still keep the `$(eval export ...)` as that should only be done for certain recipes.

camscale replied:
I've fixed that and checked it does not try to run the `security` command on linux or for other mac targets.

r0mant commented on Makefile:
```diff
@@ -429,6 +429,34 @@ build-archive:
+TSH_BUNDLEID_promote = $(TEAMID).com.gravitational.teleport.tsh
+TSH_SKELETON_promote = tsh
+
+DEVELOPER_KEY_NAME_build = Developer ID Application: Ada Lin
+INSTALLER_KEY_NAME_build = Developer ID Installer: Ada Lin
+TEAMID_build = K497G57PDJ
+TELEPORT_BUNDLEID_build = com.goteleport.dev
+TSH_BUNDLEID_build = $(TEAMID).com.goteleport.tshdev
```
Should we keep these in the Makefile, or would it make more sense to keep these as variables/secrets in Github environments (if that's a thing) or something?

r0mant replied:
Looks like they were set in the other script before so I guess it's ok.

camscale replied:
Walt did not want to put them in the github environment, and they were in the script before, so I went with that. Also putting them in the environment makes it harder to run the commands locally for testing - you still need these vars somewhere for that.

wadells reviewed (commented):
Needs the correct cert info for promote/release.  Otherwise, it looks reasonable to me.

camscale commented on build.assets/keychain-setup.sh:
It is used in https://github.com/gravitational/teleport.e/pull/1019 for CI to call to set up.

camscale replied:
I've added a bit more to the comment at the top of the keychain-setup.sh script. Let me know if you think that is sufficient.

camscale committed:
release: Split MacOS signing vars into separate mk file

Put the MacOS signing variables into a separate `.mk` file and include
it from the main `Makefile`. Add more comments to document the purpose
of the vars and where some of the values come from.

camscale committed:
release: Add some more comments to keychain-setup.sh

Explain that the purpose of the script is to be run on CI, but can also
be run manually.

Add the default values used to the usage message for the keychain and
password.

codingllama reviewed (commented):
Thanks for the heads up, Cam.

Did a quick pass and let a few comments, not a full review by any means. (Leaving the full review for assigned folks.)

> Developers can manually export ENVIRONMENT_NAME=build to use the development keys locally

Nice. Is that somewhere easy to find? Would the scripts tell me that on error?

camscale commented on build.assets/keychain-setup.sh:
```diff
@@ -0,0 +1,216 @@
+#!/usr/bin/env bash
```
I do this when things need to work on a Mac because often there is a newer version of bash in /usr/local/bin because the system one is so terribly old. But I see your point and I have been making sure not to use any bash 4+ features so yeah, I'll just use /bin/bash.

To be honest though, I don't think a malicious bash is an issue. Every command run from the script is found via the path, so any of them could be malicious too. Or is there some sort of other attack that I'm missing?

codingllama replied:
This is mostly habit/one less thing to worry about. In other contexts I wouldn't take anything from path either, but that was somewhere else.

I thought you'd mention macOS bash. Yep, it's awful.

camscale commented on build.assets/keychain-setup.sh:
```diff
@@ -0,0 +1,216 @@
+#
+#-----------------------------------------------------------------------------
+usage() {
+	printf 'Usage: %s {<options>...}\n' "${0##*/}"
+	printf 'Available options:\n'
+	printf '  -a <envvar>    take base64-encoded application key from <envvar>\n'
+	printf '  -a @<file>     take application key from <file>\n'
+	printf '  -A <envvar>    take password for application key from <envvar>\n'
+	printf '  -i <envvar>    take base64-encoded installation key from <envvar>\n'
+	printf '  -i @<file>     take installation key from <file>\n'
+	printf '  -I <envvar>    take password for installation key from <envvar>\n'
+	printf '  -k <keychain>  create <keychain>.keychain (default "build")\n'
+	printf '  -p <password>  use <password> on keychain (default "insecure")\n'
+	printf '  -v             verbose. Print commands before running them\n'
+	printf '  -n             dry run. Do not run commands, just print them\n'
```
Done

camscale commented on build.assets/keychain-setup.sh:
```diff
@@ -0,0 +1,216 @@
+
+#-----------------------------------------------------------------------------
+main() {
+	set -euo pipefail
```
I've taken to writing scripts that can also be sourced - see the bottom line of the script. This means I don't want the script to do anything if it is sourced - it should just provide functions. A `set -eu` outside of main would change the shell that sources this script, so I dont do that.

camscale commented on build.assets/keychain-setup.sh:
```diff
@@ -0,0 +1,216 @@
+	run security set-keychain-settings "${keychain}" # keep keychain unlocked
+
+	# Add the new keychain to the search path, otherwise codesign does not find the keys
+	kpath="$(security list-keychains -d user | sed 's/.*"\([^"]*\)"/\1/')"
```
Yep, shellcheck has been run. `kpath` is declared as a local at the top of the function with the other locals.

I don't see how this swallows errors though. `local foo=$(...)` does swallow errors but I'm not doing that.

I'll move the local declaration next to where it is used.

codingllama replied:
> kpath is declared as a local at the top of the function with the other locals

Sorry, I missed that.

> I'll move the local declaration next to where it is used

+1, case in point I did miss the declaration in the first read.

camscale commented:
> > Developers can manually export ENVIRONMENT_NAME=build to use the development keys locally
> 
> Nice. Is that somewhere easy to find? Would the scripts tell me that on error?

The script does not rely or use ENVIRONMENT_NAME itself. I have put some docs in the `darwin-signing.mk` file in this PR which does depend on ENVIRONMENT_NAME - let me know if you think this needs to be called out elsewhere or more detail needed.

camscale committed:
Address PR comments on keychain-setup.sh script

* Change shebang to /bin/bash
* Use heredoc instead of multiple printfs for usage message
* Move `local` declaration next to setting of kpath var

camscale committed:
release: Export DEVELOPER_ID_APPLICATION in release-darwin

The sub-make for enterprise needs this to be set or it cannot sign the
enterprise binaries. Export it if we are doing signing/notarizing.

wadells commented on darwin-signing.mk:
```diff
@@ -0,0 +1,76 @@
+
+# --- build environment
+TEAMID_build = K497G57PDJ
+DEVELOPER_KEY_NAME_build = Developer ID Application: Ada Lin
+INSTALLER_KEY_NAME_build = Developer ID Installer: Ada Lin
```
cc @adaadb6 This developer account is getting codified in our release pipeline.  It is only used for signing binaries that are pending release.  When they're ready to be promoted the signatures are stripped, and it is re-signed with the `Gravitational Inc` certificates.

camscale replied:
@wadells I'm not sure what the release process is meant to be - I was not planning on stripping any signatures and re-signing. The builds from master and branch/* would be built from source and signed with the `promote` keys. This is how the pipelines currently work. Are we changing the release process around this?

wadells reviewed (approved):
LGTM!
